src/Html/HtmlCssSelectors.fs

```diff
@@ -0,0 +1,375 @@
+                | '~' :: '=' :: t ->
+                    let s, t' = readString "" t
+                    tokenize' (AttributeValue(getOffset t + 1, s) :: AttributeContainsWord(getOffset t) :: acc) t'
+                | '!' :: '=' :: t ->
+                    let s, t' = readString "" t
+                    tokenize' (AttributeValue(getOffset t + 1, s) :: AttributeNotEqual(getOffset t) :: acc) t'
+                | StartsWith ":checkbox" t  ->
+                    tokenize' (Checkbox(getOffset t + 1) :: acc) t
+                | StartsWith ":selected" t  ->
+                    tokenize' (Selected(getOffset t + 1) :: acc) t
+                | StartsWith ":checked" t ->
+                    tokenize' (Checked(getOffset t + 1) :: acc) t
+                | StartsWith ":button" t -> 
+                    tokenize' (Button(getOffset t + 1) :: acc) t
+                | StartsWith ":hidden" t -> 
+                    tokenize' (Hidden(getOffset t + 1) :: acc) t
+                | StartsWith ":radio" t -> 
+                    tokenize' (Radio(getOffset t + 1) :: acc) t
+                | StartsWith ":password" t -> 
+                    tokenize' (Password(getOffset t + 1) :: acc) t
+                | StartsWith ":empty" t ->
+                    tokenize' (EmptyNode(getOffset t + 1) :: acc) t
+                | TokenStr ":disabled" t ->
+                    let _, t' = readString "" t
+                    tokenize' (Disabled(getOffset t + 1) :: acc) t'
+                | StartsWith ":enabled" t ->
+                    tokenize' (Enabled(getOffset t + 1) :: acc) t
+                | StartsWith ":file" t ->
+                    tokenize' (File(getOffset t + 1) :: acc) t
+                | '>' :: t ->
+                    let seqtoken = acc |> List.toSeq |> Seq.skip(1) |> Seq.toList
+                    match acc.Head with
+                        | AllChildren _ -> tokenize' (DirectChildren(getOffset t) :: seqtoken) t
+                        | _ -> tokenize' (DirectChildren(getOffset t) :: acc) t
+                | c :: t when Char.IsLetterOrDigit(c) -> 
+                    let str = c.ToString()
+                    let s, t' = readString str t
+                    tokenize' (TagName(getOffset t, s) :: acc) t'
+                | [] -> List.rev acc // TODO: refactor code to remove this
+                | c :: _ when Char.IsLetterOrDigit c |> not ->
+                    []
+                    //let offset = getOffset t
+                    //failwith (sprintf "Invalid css selector syntax (char '%c' at offset %d)" c offset)
+                | _ ->
+                    //failwith "Invalid css selector syntax"
+                    []
+            tokenize' [] source
+
+    type FilterLevel = 
+        | Root
+        | Children
+        | Descendants
+
+    type CssSelectorExecutor(nodes:HtmlNode list, tokens:SelectorToken list) = 
+        let mutable level = FilterLevel.Descendants
```

Comment:
You could pass  ```level```  in recursion, and use it as an argument to ```getTargets``` and other functions.
This would make th intent of level simpller and probably remove the need for a class.